File: online-bookstore/src/app/service/book.service.ts
import { Injectable } from '@angular/core';
import { Book } from '../model/book.module';
import { Router } from '@angular/router';
import { Observable, tap } from 'rxjs';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class BookService {
 

  navigateToLink(arg0: string) {
    this.router.navigate(['/'+ arg0]);
  }
 
  private baseUrl = 'http://localhost:9005/book';
  private URL = 'http://localhost:9005';

  constructor(private http: HttpClient,private router:Router) { }

  getAllCategory(): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.URL}/categories/veiwcategory`);
  }

  // Get all books
  getAllBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/viewbooks`);
  }

  // Add a new book
  addBook(body: any): Observable<any> {
    return this.http.post(`${this.baseUrl}/addbook`, body,{ responseType: 'text' });
    
  }

  // Update an existing book
  updateBook(id: number, book: Book): Observable<Book> {
    return this.http.put<Book>(`${this.baseUrl}/updatebook/${id}`, book).pipe(
      tap((updatedBook: Book) => console.log(`Updated book with ID=${updatedBook.id}`))
    );
  }

  // Remove a book by ID
  removeBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/removeBook/${id}`, { responseType: 'text' as 'json' }).pipe(
      tap(() => console.log(`Deleted book with ID=${id}`))
    );
  }

  // Find a book by ID
  findBookById(id: number): Observable<Book> {
    return this.http.get<Book>(`${this.baseUrl}/findbookbyid/${id}`);
  }

  // Find books by name
  findByBookname(bookname: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/findbybookname/${bookname}`);
  }

  // Find books by category
  findByCategory(category: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/findbycategory/${category}`);
  }

  // Find books by ISBN
  findByIsbn(isbn: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/findbyisbn/${isbn}`);
  }

  // Find books by author
  findByAuthor(author: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/findbyauthor/${author}`);
  }

  // Find books by publisher
  findByPublisher(publisher: string): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/findbypublisher/${publisher}`);
  }
  searchBooks(searchText: string): Observable<Book[]>{
    return this.http.get<Book[]>(`${this.baseUrl}/search?searchText=${searchText}`);
  }
  getBookCount(): Observable<number> {
    console.log('call');
    return this.http.get<number>(this.baseUrl+'/count');
  }
}

File: bookfinalproject/src/main/java/com/spring/service/BookService.java
package com.spring.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.spring.entity.Book;
import com.spring.entity.Category;
import com.spring.repository.BookRepository;

@Service
public class BookService {
	@Autowired
    private BookRepository bookRepository;
	
	 public List<Book> getAll(){
		 return bookRepository.findAll();
	 }
	 
	 public Book addBook(Book b) {
		 return bookRepository.save(b);
	 }
	 
	 public Book updateBook(Long id,Book b) {
		 b.setId(id);
		 return bookRepository.save(b);
	 }
	 
	 public void removeBook(long id) {
		 bookRepository.deleteById(id);
	 }
	 
	 public Book findBookById(long id) {
		 return bookRepository.findById(id).orElse(null);
	 }
	 
	 public List<Book> findByBookname(String bookname){
		 return bookRepository.findByBookname(bookname);
	 }
	 
	 public List<Book> findByCategory(Category category){
		 return bookRepository.findByCategory(category);
	 }
	 
	 public List<Book> findByisbn(String isbn){
		 return bookRepository.findByisbn(isbn);
	 }
	 
	 public List<Book> findByAuthor(String author){
		 return bookRepository.findByAuthor(author);
	 }
	 
	 public List<Book> findByPublisher(String publisher){
		 return bookRepository.findByPublisher(publisher);
	 }
	 public long getBookCount() {
	        return bookRepository.countBooks();
	    }
}
